xtask/src/tools/lychee.rs

```diff
@@ -0,0 +1,128 @@
+};
+use saucer::{anyhow, Error, Result, Utf8PathBuf};
+
+use crate::utils::PKG_PROJECT_ROOT;
+
+pub(crate) struct LycheeRunner {
+    client: Client,
+    verbose: bool,
+}
+
+impl LycheeRunner {
+    pub(crate) fn new(verbose: bool) -> Result<Self> {
+        let accepted = Some(HashSet::from_iter(vec![
+            StatusCode::OK,
+            StatusCode::TOO_MANY_REQUESTS,
+        ]));
+
+        let client = ClientBuilder::builder()
+            .exclude_all_private(true)
+            .retry_wait_time(Duration::from_secs(30))
+            .max_retries(5u8)
+            .accepted(accepted)
+            .build()
+            .client()?;
+
+        Ok(Self { client, verbose })
+    }
+
+    pub(crate) fn lint(&self) -> Result<()> {
+        if self.verbose {
+            println!("Checking links in documentation");
+        }
+
+        let inputs: Vec<Input> = get_md_files()
+            .iter()
+            .map(|file| Input {
+                source: InputSource::FsPath(PathBuf::from(file)),
+                file_type_hint: Some(FileType::Markdown),
+                excluded_paths: None,
+            })
+            .collect();
+
+        let rt = Runtime::new()?;
+
+        let lychee_client = self.client.clone();
+
+        rt.block_on(async move {
+            let links = Collector::new(None)
+                .collect_links(inputs)
+                .await
+                .collect::<LycheeResult<Vec<_>>>()
+                .await?;
+
+            let mut has_failures = false;
+            let links_size = links.len();
+
+            for link in links {
+                let response = lychee_client.check(link).await?;
+                if response.status().is_failure() {
+                    has_failures = true;
+                    if self.verbose {
+                        println!("[x] {}", response.1.uri.as_str());
+                    }
+                } else if response.status().is_success() {
+                    println!("[✓] {}", response.1.uri.as_str());
+                }
+            }
+
+            println!("{} links checked.", links_size);
+
+            if has_failures {
+                return Err(anyhow!("Some links in markdown documentation are down."));
+            }
+
+            Ok::<(), Error>(())
+        })?;
+
+        Ok(())
+    }
+}
+
+fn get_md_files() -> Vec<Utf8PathBuf> {
+    let mut md_files = Vec::new();
+
+    walk_dir(PKG_PROJECT_ROOT.as_str(), &mut md_files);
+
+    md_files
+}
+
+fn walk_dir(base_dir: &str, md_files: &mut Vec<Utf8PathBuf>) {
+    if let Ok(entries) = fs::read_dir(base_dir) {
+        for entry in entries.flatten() {
+            if let Ok(file_type) = entry.file_type() {
+                if file_type.is_file() {
+                    if let Ok(file_name) = entry.file_name().into_string() {
+                        // the CHANGELOG is simply too large to be running this check on every PR
+                        if file_name.ends_with(".md") && !file_name.contains("CHANGELOG") {
```

Comment:
i'm interested to see if `lychee` can handle this better than our existing link checker. do you mind trying to run it and see if it works OK? there are a _lot_ of links in that file so it might still be worth skipping, but i want to see if rust makes it more manageable.

Reply:
[without CHANGELOG](https://app.circleci.com/pipelines/github/apollographql/rover/1813/workflows/f9551c55-e352-4b09-b49f-73a88ad47afd/jobs/11062), it takes ~51s vs ~1min25 [with](https://app.circleci.com/pipelines/github/apollographql/rover/1814/workflows/24871c94-bed4-4b8a-8844-ded386a69f88/jobs/11068)

I think there is also a bug because it detects `harmonizer@0.2.2` as a link 🤔 .
Anyway, I think this acceptable. Let me check with the old tool.

Reply:
oh yeah that's definitely worth it! let's not exclude that then 😄